components/gitpod-cli/cmd/credential-helper.go

```diff
@@ -41,42 +55,82 @@ var credentialHelper = &cobra.Command{
 			if token == "" {
 				token = "no"
 			}
+			log.Infof("returning username=%s\npassword=%s\n", user, token)
 			fmt.Printf("username=%s\npassword=%s\n", user, token)
 		}()
 
-		log.SetOutput(io.Discard)
-		f, err := os.OpenFile(os.TempDir()+"/gitpod-git-credential-helper.log", os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
-		if err == nil {
-			defer f.Close()
-			log.SetOutput(f)
-		}
-
-		url, gitCommand := parsePstree()
+		repoURL, gitCommand := parsePstree()
 		host := parseHostFromStdin()
 		if len(host) == 0 {
 			log.Println("'host' is missing")
 		}
 
-		service, err := theialib.NewServiceFromEnv()
+		if isTheiaIDE() {
+			service, err := theialib.NewServiceFromEnv()
+			if err != nil {
+				log.WithError(err).Print("cannot connect to Theia")
+				return
+			}
+			if action == "get" {
+				resp, err := service.GetGitToken(theialib.GetGitTokenRequest{
+					Command: gitCommand,
+					Host:    host,
+					RepoURL: repoURL,
+				})
+				if err != nil {
+					log.WithError(err).Print("cannot get token")
+					return
+				}
+				user = resp.User
+				token = resp.Token
+			}
+			return
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+		defer cancel()
+		supervisorAddr := os.Getenv("SUPERVISOR_ADDR")
+		if supervisorAddr == "" {
+			supervisorAddr = "localhost:22999"
+		}
+		supervisorConn, err := grpc.Dial(supervisorAddr, grpc.WithInsecure())
 		if err != nil {
-			log.WithError(err).Print("cannot connect to Theia")
+			log.WithError(err).Print("error connecting to supervisor")
 			return
 		}
-
-		resp, err := service.GetGitToken(theialib.GetGitTokenRequest{
-			Command: gitCommand,
-			Host:    host,
-			RepoURL: url,
+		resp, err := supervisor.NewTokenServiceClient(supervisorConn).GetToken(ctx, &supervisor.GetTokenRequest{
+			Host: host,
+			Kind: "git",
 		})
 		if err != nil {
-			log.WithError(err).Print("cannot get token")
+			log.WithError(err).Print("error getting token from supervisior")
+			return
+		}
+
+		log.Infof("Runnig gp git-token-validator --user %s --token %s --scopes %s --host %s --repoURL %s --gitCommand %s",
+			resp.User, resp.Token, strings.Join(resp.Scope, ","), host, repoURL, gitCommand)
+		validator := exec.Command("gp", "git-token-validator",
+			"--user", resp.User, "--token", resp.Token, "--scopes", strings.Join(resp.Scope, ","),
+			"--host", host, "--repoURL", repoURL, "--gitCommand", gitCommand)
+		err = validator.Start()
```

Comment:
Considering that `git-token-validator` is also part of the gitpod-cli, why not wrap that in a package and call it in-process? It feels odd that we have to call a separate program here.

If we need to do it this way, then we should be calling `/proc/self/exe` rather than `gp` because it doesn't make the assumption that `gp` exists in the path and points to this executable.

Reply:
Git expects the credential helper to exit, that's why.  Switching to /prog/self/exe